Change to using `ramsey/composer-install`

desrosj opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/53841

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -71,12 +71,21 @@ jobs:
         run: |
-          php --version
           composer --version
+          php --version
+          php -i
```
While I can see `php -i` _may_ be of some use, the `php --version` command seems redundant as this information is already available in the fold-out from `shivammathur/setup-php`.

Actually, same goes for `composer --version`, again, that's already available in the `setup-php` fold-out.

desrosj replied:
Agreed that this is redundant.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -71,12 +71,21 @@ jobs:
+          dependency-versions: 'highest'
+
+      - name: Make Composer packages available globally
+        run: echo "${PWD}/vendor/bin" >> $GITHUB_PATH
```
I don't understand why you are changing the order of the steps.

I mean, the `ramsey/composer-install` action doesn't install Composer.
The standard `ubuntu-latest` image comes with PHP and Composer pre-installed in the first place and then when using the `shivammathur/setup-php` action, `setup-php` will install a specific PHP version and will make sure the Composer version installed is the most up to date version for the PHP version selected.

desrosj replied:
Ah, thanks for pointing this out. When I was thinking "group setup tasks first". But you're right that this only installs dependencies and not Composer itself. Moved them back.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -90,14 +99,6 @@ jobs:
-        run: echo "${PWD}/vendor/bin" >> $GITHUB_PATH
-
       - name: Log PHPCS debug information
         run: phpcs -i
```
Along the same lines as my other note about redundant information logging - the output of this command will already be displayed at the bottom of the Composer `composer install/update` run (generated by the `dealerdirect/phpcodesniffer-composer-installer` plugin).

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -71,12 +71,21 @@ jobs:
+      - name: Install Composer dependencies
+        uses: ramsey/composer-install@f680dac46551dffb2234a240d65ae806c2999dd6 # v2.1.0
+        with:
+          dependency-versions: 'highest'
```
This is not actually needed. WP does not have a committed `composer.lock` file, so `ramsey/composer-install` will automatically run `composer update` instead of `composer install`.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -71,12 +71,21 @@ jobs:
           php-version: '7.4'
           coverage: none
-          tools: composer, cs2pr
+          tools: cs2pr
```
This is correct. The `composer` in the `tools` key was never needed (unless you want to install a specific version of Composer or want to uninstall it, but neither is the case).
`shivammathur/setup-php` will always install Composer automatically. No need to request it.

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -114,37 +114,38 @@ jobs:
-          fi
+      - name: General debug information
+        run: |
+          npm --version
+          node --version
```
The latest version of `actions/setup-node` - `3.5.1`, which came out today, will automatically show this information already.

See: https://github.com/actions/setup-node/releases/tag/v3.5.1

desrosj replied:
`setup-node` does display this information, but it hides it in a folded section labelled "Environment details", which is not really clear. I think leaving this one will make it much easier to find this information compared to `setup-php`, which shows the information without having to expand any sections.

jrfnl replied:
Makes sense to me.

jrfnl reviewed (commented):
Hi @desrosj, thanks for setting this up!

I've left a number of remarks inline.

Same comments as I left for the CS workflow apply to the PHPCompatibility workflow (and in parts also to the other workflows).

Regarding my comments about redundant debug logging - I'm not saying that it should be removed, just wanted to point out that most of this information is already available.
Not everyone is aware that it is, so wanted to make sure that this information is known.

desrosj commented:
I updated the debug data so it no longer is printed before _and_ after configuring PHP, Node.js, etc. We only care about logging the versions _after_ these steps. But now we're straying a bit from the original intent of the PR. I'll consider splitting out the debug information improvements into a separate commit once we're happy with everything here.

desrosj commented:
@jrf any objection to merging this now with the intention of following up in 2 weeks to either update to a new version of `install-composer` fixing this issue? Or perhaps using a custom cache key here temporarily until the update becomes available?

jrfnl commented on .github/workflows/end-to-end-tests.yml:
```diff
@@ -61,23 +60,21 @@ jobs:
           curl --version
           git --version
           svn --version
-          php --version
-          php -i
```
This workflow doesn't use `setup-php`, so logging the PHP information here does seem useful. Then again, there is also the Docker debug info (on line 91) which is probably more pertinent, so yeah, let's remove this here as it could well confuse people if the PHP version mentioned here vs the one mentioned in the Docker debug is different.

(i.e. no action needed, just documenting the thought process here for future reference)

jrfnl reviewed (commented):
I agree, splitting this up into two commits makes sense:
1. Handling the debug logging changes (and moving the node install up to make sure the debug logs contains the correct information).
2. Handling the change over to the `composer-install` action runner.

In my opinion, the logging changes can go in straight away.

As for the change to `composer-install` - I'd favour waiting a week or two for the following reasons:
1. It doesn't solve the current `set-output` warnings issue until a new release has been tagged.
2. For the PHP test runs, the cache key/cache refreshing would be worse than it currently is, until the `custom-cache-suffix` feature has been merged & released (in progress).
3. It would create the need for a follow-up commit soon after.

As a general question - may just be me -, but why does each workflow have a "Performs the following steps" docblock ?
The steps all have names, and if needs be, could be accompanied by a comment _within_ the workflow (above the step name).

The step docs at the top feel (to me) like something which can very easily get out of sync with the reality, making it a maintenance burden we can do without.

desrosj commented:
> As a general question - may just be me -, but why does each workflow have a "Performs the following steps" docblock ? The steps all have names, and if needs be, could be accompanied by a comment _within_ the workflow (above the step name).
> 
> The step docs at the top feel (to me) like something which can very easily get out of sync with the reality, making it a maintenance burden we can do without.

I think that initially, there were some steps that required additional context that was a little long for a `name` field, so I included it in this docblock.

In most cases, I think the need for extra context has disappeared. There could be scenarios in the future where it may be valuable, but It is a bit redundant. I wonder if these docblocks make it easier to follow and understand what is happening for anyone newer to GitHub Actions. I don't feel strongly either way.

desrosj commented:
Today I discovered that there are [cache related endpoints in the REST API now for GitHub Actions](https://docs.github.com/en/rest/actions/cache#about-the-cache-api).

Having a workflow that periodically deletes certain cache entries could be added to solve the problem with cache keys entirely. An upstream way to handle this would still be nice, but I'm thinking there may be other times we want to clear some caches from the repository to optimize our usage of caching with actions.

One that comes to mind that I noticed recently looking at the caches stored on `wordpress-develop` was tags. The hashes for older branches probably don't need to be stored past a day or two since they don't have a lot of activity. Clearing these out would allow other cache keys with a higher likelihood of being reused to persist longer. Currently, it seems that the oldest cache entry on `wordpress-develop` is 3 days.

Definitely something to explore in its own ticket and discussion, but wanted to mention that here.

jrfnl commented:
Also: just saw that this was published yesterday: [GH Blog: Manage caches in your Actions workflows from Web Interface](https://github.blog/changelog/2022-10-20-manage-caches-in-your-actions-workflows-from-web-interface/) ;-)

desrosj committed:
Remove `"--no-progress --no-ansi"` options for `composer-install` action.

These are now applied by default.

desrosj committed:
Use `ramsey/composer-install` in all workflows using Composer.

desrosj committed:
Don't use `setup-php` to install Composer.

`composer-install` handles this.

desrosj committed:
Implement `custom-cache-suffix`.

This ensures that the Composer cache is flushed at least once per week, which ensures the cache is relatively up to date with the results of `composer update`.

desrosj commented:
I've gone and cleaned up the PR after committing most of the changes unrelated to the goal of this PR/Trac ticket separately in [Trac-54851](https://core.trac.wordpress.org/changeset/54851/) and implemented the new `custom-cache-suffix` feature from upstream.

desrosj committed:
Fix bad search/replace.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -86,10 +85,12 @@ jobs:
         uses: ramsey/composer-install@83af392bf5f031813d25e6fe4cd626cdba9a2df6 # v2.2.0
         with:
-          composer-options: "--no-progress --no-ansi"
+          custom-cache-suffix: $(/bin/date -u --date='last Mon' "+%F")
```
Considering the date is still retrieved in a separate step anyway for the PHPCS cache, why not re-use it ?

```suggestion
          custom-cache-suffix: ${{ steps.get-date.outputs.date }}
```

As a side-note: what with the two PHPCS runs in this workflow and these not using separate PHPCS caches, the caching of the PHPCS cache file is kind of useles....

desrosj replied:
I did some testing and these were my findings:

Scenario 1: Run `phpcs -n` first
- first run: 28.41s
- second run: 1.56s
- first run of `phpcs tests`: 678ms
- second run of `phpcs tests`: 533ms
- rerunning `phpcs -n`: 879ms

Scenario 2: Run `phpcs tests` first
- first run: 16.16s
- second run: 331ms
- first run of `phpcs -n`: 13.4s
- second run of `phpcs -n`: 1.43s
- rerunning `phpcs tests`: 403ms

I'm not familiar with how PHPCS generates the cache file, but it seems that there's still a benefit to caching `.cache/phpcs.json`, even if the cache file does not include _everything_ we would like to be cached.

Let's open a different ticket to discuss further if we want to perform a deeper dive.

jrfnl replied:
Would you mind opening the issue for it ? I can send in a PR to fix it. We'd only need to use different cache file names to get the optimal benefit in CI and that can be sorted via CLI arguments.

desrosj replied:
Done! https://core.trac.wordpress.org/ticket/57148

jrfnl commented on .github/workflows/php-compatibility.yml:
```diff
@@ -84,10 +84,12 @@ jobs:
         uses: ramsey/composer-install@83af392bf5f031813d25e6fe4cd626cdba9a2df6 # v2.2.0
         with:
-          composer-options: "--no-progress --no-ansi"
+          custom-cache-suffix: $(/bin/date -u --date='last Mon' "+%F")
```
Same as above (though this workflow doesn't have the PHPCS caching issue as there is only one PHPCS run).

```suggestion
          custom-cache-suffix: ${{ steps.get-date.outputs.date }}
```

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -112,45 +112,39 @@ jobs:
+        uses: ramsey/composer-install@83af392bf5f031813d25e6fe4cd626cdba9a2df6 # v2.2.0
+        with:
+          custom-cache-suffix: $(/bin/date -u --date='last Mon' "+%F")
+          composer-options: "${{ matrix.php == '8.2' && '--ignore-platform-req=php+' || '' }}"
```
`--ignore-platform-req=php+` is no longer needed for PHP 8.2 AFAICS, so the whole `composer-options` line can be removed.

jrfnl reviewed (changes requested):
Left a few small remarks, other than that, looks good to me!

desrosj committed:
Reuse previously calculated date.

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

desrosj committed:
Reuse previously calculated date.

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

desrosj committed:
Remove `composer-options`.

The `--ignore-platform-req=php+` is no longer required for PHP 8.2.

desrosj commented:
Merged in https://core.trac.wordpress.org/changeset/54856.

jrfnl commented:
Nice! Glad it's sorted now!
